Source/WebCore/platform/audio/cocoa/AudioSampleBufferList.cpp

```diff
@@ -112,40 +110,42 @@ void AudioSampleBufferList::applyGain(float gain)
 
 static void mixBuffers(WebAudioBufferList& destinationBuffer, const AudioBufferList& sourceBuffer, AudioStreamDescription::PCMFormat format, size_t frameCount)
 {
-    for (uint32_t i = 0; i < destinationBuffer.bufferCount(); i++) {
+    auto sourceBufferSpan = span(sourceBuffer);
+    auto destinationBufferSpan = span(*destinationBuffer.list());
+    for (auto [source, destination] : zippedRange(sourceBufferSpan, destinationBufferSpan)) {
         switch (format) {
         case AudioStreamDescription::Int16: {
-            ASSERT(frameCount <= destinationBuffer.buffer(i)->mDataByteSize / 2);
-            ASSERT(frameCount <= sourceBuffer.mBuffers[i].mDataByteSize / 2);
+            ASSERT(frameCount <= destination.mDataByteSize / 2);
+            ASSERT(frameCount <= source.mDataByteSize / 2);
 
-            auto* destination = static_cast<int16_t*>(destinationBuffer.buffer(i)->mData);
-            auto* source = static_cast<int16_t*>(sourceBuffer.mBuffers[i].mData);
-            for (size_t i = 0; i < frameCount; i++)
-                destination[i] += source[i];
+            auto sourceData = dataSpan<int16_t>(source);
+            auto destinationData = dataMutableSpan<int16_t>(destination);
+            for (auto [s, d] : zippedRange(sourceData, destinationData))
+                d += s;
             break;
         }
         case AudioStreamDescription::Int32: {
-            ASSERT(frameCount <= destinationBuffer.buffer(i)->mDataByteSize / 4);
-            ASSERT(frameCount <= sourceBuffer.mBuffers[i].mDataByteSize / 4);
+            ASSERT(frameCount <= destination.mDataByteSize / 4);
+            ASSERT(frameCount <= source.mDataByteSize / 4);
 
-            auto* destination = static_cast<int32_t*>(destinationBuffer.buffer(i)->mData);
-            vDSP_vaddi(destination, 1, reinterpret_cast<int32_t*>(sourceBuffer.mBuffers[i].mData), 1, destination, 1, frameCount);
+            auto* destinationPtr = static_cast<int32_t*>(destination.mData);
+            vDSP_vaddi(destinationPtr, 1, reinterpret_cast<int32_t*>(source.mData), 1, destinationPtr, 1, frameCount);
```

Comment:
Technically this is a violation of bounds safety -- though it is our callee that does the unbounded dereference, rather than our own code. We have ASSERTs, but those don't do anything in release builds.

The most principled thing to do would be to create wrappers like `vDSP_vaddi_span` that took spans instead of pointers, and RELEASE_ASSERTed that all three spans were big enough for frameCount before calling through to vDSP_vaddi. That's the moral equivalent of what we do with memcpySpan.

Reply:
OK, I'll add a wrapper like so to VectorMath.cpp:
```
void add(std::span<const int> inputVector1, std::span<const int> inputVector2, std::span<int> outputVector)
{
    RELEASE_ASSERT(inputVector1.size() == inputVector2.size());
    RELEASE_ASSERT(outputVector.size() >= inputVector1.size());
    vDSP_vaddi(inputVector1.data(), 1, inputVector2.data(), 1, outputVector.data(), 1, inputVector1.size());
}
```